File: Stack/Python/150_reverse_polish_notation.py
# RUNTIME: 3 ms, faster than 63.94% of python3 Submissions. 
# MEMORY USAGE: 19.06 MB, less than 87.91% of python3 Submissions.


# My thougtprocess:
# 1. Define a open stack
# 2. Define operations - what happens when "+" ist in list etc.
# 3. Check if character is in operation list
# 4. If not simply append
# 5. Else use predetermined operation with two right most values
# 6. If deviding need to cast int or we may get float value
# 7. Return remaining stack element.


import operator
from typing import List



tkns = ["10","6","9","3","+","-11","*","/","*","17","+","5","+"]

class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        st = []
        ops = {
            "+": operator.add,
            "-": operator.sub,
            "*": operator.mul,
            "/": lambda left, right: int(left / right)
        }
        for c in tokens:
            if c not in ops:
                st.append(int(c))
            else:
                right = st.pop()
                left = st.pop()
                st.append(ops[c](left, right))
        return st[-1]





sol = Solution()
print(sol.evalRPN(tkns))  # 22

File: Stack/C++/150_reverse_polish_notation.cpp
/*
* RUNTIME: 0 ms - Beats 100.00% of submissions  - don´t know how 0 ms is even possible
* MEMORY - 17.37 MB - Beats 11.74% of submissions
*/

#include <vector>
#include <string>
#include <algorithm>
#include <set>
#include <stack>
#include <unordered_map>
#include <functional>
using namespace std;

/* Thoughtprocess:
* 1. Define a stack to store operands
* 2. Define operations - what happens when "+" is in the list, etc.
* 3. Check if the current token is in the operations map
* 4. If not, simply push the number onto the stack
* 5. Else, apply the predefined operation using the two most recent values
* 7. Return the remaining element on the stack
*
*/


class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        stack<int> st;
        
        unordered_map<string, function<int(int,int)>> ops = 
        {
            {"+", [](int a, int b) { return a + b; }},
            {"-", [](int a, int b) { return a - b; }},
            {"*", [](int a, int b) { return a * b; }},
            {"/", [](int a, int b) { return a / b; }}
        };
        
        for (string s : tokens)
        {
            if ( ops.find(s) == ops.end())
            {
                st.push(stoi(s));
            }
            else
            {
                int right = st.top();
                st.pop();
                int left = st.top();
                st.pop();
                st.push(ops.find(s)->second(left, right));
            }
        }
        return st.top();
    }
};